src/rules/no_undef.rs

```diff
@@ -26,139 +23,809 @@ impl LintRule for NoUndef {
+    //   let finder = decl_finder(src);
+    //   let target_call = &idents[1];
+    //   assert!(!finder.decl_exists(&target_call));
+    // }
+
+    #[test]
+    fn function_call_in_export_default_self() {
+      let src = r#"
+export default function target() {
+  target();
+}
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_call = &idents[1];
+      assert!(finder.decl_exists(&target_call));
+    }
+
+    #[test]
+    fn decl_in_child_scope() {
+      let src = r#"
+function foo() {
+  let target = 0;
+}
+target = 1;
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_assignment = &idents[1];
+      assert!(!finder.decl_exists(&target_assignment));
+    }
+
+    #[test]
+    fn decl_as_default_import() {
+      let src = r#"
+import target from "mod.ts";
+target();
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_call = &idents[1];
+      assert!(finder.decl_exists(&target_call));
+    }
+
+    #[test]
+    fn decl_as_named_import() {
+      let src = r#"
+import { target } from "mod.ts";
+target();
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_call = &idents[1];
+      assert!(finder.decl_exists(&target_call));
+    }
+
+    #[test]
+    fn decl_as_star_as_import() {
+      let src = r#"
+import * as target from "mod.ts";
+target();
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_call = &idents[1];
+      assert!(finder.decl_exists(&target_call));
+    }
+
+    #[test]
+    fn decl_enum() {
+      let src = r#"
+enum Target {
+  Foo,
+  Bar,
+}
+const a = Target.Foo;
+        "#;
+      let idents = get_idents(src, "Target");
+      let finder = decl_finder(src);
+      let target_used = &idents[1];
+      assert!(finder.decl_exists(&target_used));
+    }
+
+    #[test]
+    fn decl_in_param_of_anonymous_function() {
+      let src = r#"
+[1,2,3].forEach(target => {
+  console.log(target);
+});
+        "#;
+      let idents = get_idents(src, "target");
+      let finder = decl_finder(src);
+      let target_used = &idents[1];
+      assert!(finder.decl_exists(&target_used));
+    }
+
+    // TODO(magurotuna): This should be passed. Variables declared with `var` are not block-scoped,
+    // but function-scoped. We need to fix the handling for `var` declared variables.
+    // #[test]
+    // fn function_scoped_var() {
+    //   let src = r#"
+    //     function foo() {
+    //       target++;
+    //       {
+    //         var target = 0;
+    //       }
+    //     }
+    //     "#;
+    //   let idents = get_idents(src, "target");
+    //   let finder = decl_finder(src);
+    //   let target_used = &idents[0];
+    //   assert!(finder.decl_exists(&target_used));
+    // }
```

Comment:
This example shows that `DeclFinder` (which is `no-undef` specific analyzer) cannot deal with `var`. If we uncomment this test, test will fail.
`var target = 0;` is not block-scoped but function-scoped, so `target++` should be treated as valid.

CC @bartlomieju @kdy1

Reply:
This seems like a bug of resolver in swc. I'll check if it's the case.

Reply:
Oh really? I think there's a problem with my implementation of `DeclFinder`. In `DeclFinder`, it collects variable declarations in a naive way. It doen't use swc's `SyntaxContext` or something like that.

Reply:
To be honest I have no idea what `SyntaxContext` exactly is. In the original implementation of `no-undef`, the following code resulted in false positive, which was perhaps caused by `SyntaxContext` problem, I guess.

```ts
(() => {
  function foo() {
    return new Bar();
  }
  class Bar {}
})();
```

Reply:
It was not, while the code in #463 is bug of the swc resolver.
I'll investigate further.

Reply:
I see. Thanks for your help :)

Reply:
Does deno_lint does some transforms?
Root cause of the failing test is `visit_var_declarator` not being called.

Reply:
Oh, I did wrong test.

Reply:
Changing to
```rust

impl BindingCollector {
  fn declare(&mut self, i: Id) {
    dbg!(&i);
    self.declared.insert(i);
  }
}

impl Visit for BindingCollector {
  fn visit_fn_decl(&mut self, f: &FnDecl, _: &dyn Node) {
    dbg!(&f.ident);
    self.declare(f.ident.to_id());
    f.visit_children_with(self);
  }
}
```

fixes the issue. (I changed it to visit children nodes)

Reply:
Great! I'll give it a try

Reply:
hmm... tried your fix but the test still fails.

Code fix:
https://github.com/magurotuna/deno_lint/blob/test-magurotuna/src/rules/no_undef.rs#L56-L66

Test case:
https://github.com/magurotuna/deno_lint/blob/test-magurotuna/src/rules/no_undef.rs#L352-L368

Test result:

```
thread 'rules::no_undef::tests::no_undef_valid' panicked at 'Unexpected diagnostics found:
[
    LintDiagnostic {
        range: Range {
            start: Position {
                line: 4,
                col: 19,
                byte_pos: 56,
            },
            end: Position {
                line: 4,
                col: 22,
                byte_pos: 59,
            },
        },
        filename: "deno_lint_test.tsx",
        message: "Bar is not defined",
        code: "no-undef",
        hint: None,
    },
]

source:

    (() => {
      function foo() {
        return new Bar();
      }
      class Bar {}
    })();

', src/test_util.rs:228:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Do I miss something?

Reply:
```js
(() => {
      function foo() {
        return new Bar();
      }
      class Bar {}
    })();
```

This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.

Reply:
> This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.This is bug of swc resolver. I'm currently working on it. I think the test below is a same bug.

Oh I see. Then I'll wait for swc resolve to be fixed 👍

Reply:
I fixed it and published a new version of `swc_ecmascript` and `swc_ecma_transforms`.

Reply:
@kdy1 
Thanks for the quick fix!
`deno_lint` depends on `swc_ecmascript`, doesn't depend on `swc_scma_transforms` directly. It seems like `swc_ecma_transforms` as a dependency of `swc_ecmascript` was not updated. (in https://github.com/swc-project/swc/blob/master/ecmascript/Cargo.toml#L27, the version of transforms is still `0.29.1`)
Is this correct?

Reply:
Oh. I forgot updating it but you can do `cargo update -p swc_ecma_transforms`.

Reply:
nice command! That allows to pass the test. 
I have one more question. Now we've added `f.visit_children_with(self);` to `visit_fn_decl`. Should we add `visit_children_with` to the other visit methods e.g. `visit_class_decl`? I feel like this is required to handle a case like the following:

```ts
class SomeClass {
  someMethod() {
    function foo() {
      return new Bar();
    }
    class Bar {}
  }
}
```

Reply:
Yes. The binding collector should collect all bindings.
Alternatively, you can change it to `VisitAll`.

Reply:
oh, right, `VisitAll` is the finest solution here. I'll use that :)